### PR TITLE
indexer, op-service: refactor service lifecycle to start/stop resoures more cleanly

### PR DESCRIPTION
**Pull request by protolambda**
**Description**

This refactors the indexer service lifecycle, to start/stop resources more cleanly.

Changes:
- Adopt the `cliapp.Lifecycle` from the op-service, which handles all the interrupt handling and error / exit-code reporting.
- Remove the "start with context, run till context shutdown" pattern: this way we can reason about running processes, and close them explicitly, while waiting for it to complete. This should help a lot against CI flakes.
- Minor change to the ETLs and bridge-processor, to move the panic-catching and shutdown-request into the service, so it can be tested, and t.Fatalf can be called as necessary.
- Add "DBConnector" to API service setup, so we can swap how the DB view is retrieved. This way we can insert the already known DB view nicely in e2e testing.
- Close a lot of resources, explicitly, and propagate the errors, so we don't have any lingering background process work when the service is closed.
- Use `clock.LoopFn` for the inner ETL functionality: we can mock this clock if necessary for testing, and it has a nice way of closing the ticking loop.

And attach the interrupt-blocker to the app context, so each app-action can block from a common continuous interrupt-channel, to not miss any interrupts, and avoid conflicts between sub-commands that may have their own life-cycles.

TLDR: No indexer functionality changes, just a lot of plumping updates in the service code to make it flake less in CI and gracefully shut down all its resources, without dropping any errors.

**Tests**

Existing e2e / unit tests cover the same functionality. Now that the service setup/shutdown is more testable I'm happy to help add e2e tests, suggestions welcome.

### CODE REVIEW

File: indexer/api/api.go [hunk 1]
```diff
@@ -45,6 +36,23 @@ const (
...
+
+	bv      database.BridgeTransfersView
+	dbClose func() error
+
+	metricsRegistry *prometheus.Registry
+
+	apiServer     *httputil.HTTPServer
+	metricsServer *httputil.HTTPServer
+
+	stopped atomic.Bool
```
Comment (roninjin10) [inline indexer/api/api.go:54]: The docs for atomic package sound like we aren't supposed to be using it:

> Package atomic provides low-level atomic memory primitives useful for implementing synchronization algorithms.
> 
> These functions require great care to be used correctly. Except for special, low-level applications, synchronization is better done with channels or the facilities of the [sync](https://pkg.go.dev/sync) package. Share memory by communicating; don't communicate by sharing memory.

**Review by roninjin10 — COMMENTED** [review]
Nice! Since I copied much of this code over to mocktimism I'll make a ticket to update mocktimism similarly

File: indexer/api/api.go (hunk 1 above)
Comment (protolambda) [reply indexer/api/api.go:54]: It's a generic warning, and they're right about communication, but we don't want data-communication with channels here, it's just a basic variable I want to set to track the state

File: indexer/api/api.go [hunk 2]
```diff
@@ -53,112 +61,116 @@ func chiMetricsMiddleware(rec metrics.HTTPRecorder) func(http.Handler) http.Hand
...
+	a.initRouter()
+	if err := a.startServer(cfg.HTTPServer); err != nil {
+		return fmt.Errorf("failed to start API server: %w", err)
+	}
+	return nil
+}
 
-	return &API{log: logger, router: apiRouter, metricsRegistry: mr, serverConfig: serverConfig, metricsConfig: metricsConfig}
+func (a *APIService) Start(ctx context.Context) error {
+	// Completed all setup-up jobs at init-time already,
```
Comment (hamdiallam) [inline indexer/api/api.go:88]: Feels a bit nicer to have `startMetricsServer` and `startServer` in `Start` whilst the db and router are bootstrapped from config

Comment (hamdiallam) [reply indexer/api/api.go:88]: perhaps hold reference to the config? Dont feel too strongly about this

File: indexer/etl/l2_etl.go [hunk 3]
```diff
@@ -68,62 +79,101 @@ func NewL2ETL(cfg Config, log log.Logger, db *database.DB, metrics Metricer, cli
...
+		return fmt.Errorf("failed to start internal ETL: %w", err)
+	}
 
-			l2ContractEvents := make([]database.L2ContractEvent, len(batch.Logs))
-			for i := range batch.Logs {
-				timestamp := batch.HeaderMap[batch.Logs[i].BlockHash].Time
-				l2ContractEvents[i] = database.L2ContractEvent{ContractEvent: database.ContractEventFromLog(&batch.Logs[i], timestamp)}
-				l2Etl.ETL.metrics.RecordIndexedLog(batch.Logs[i].Address)
+	// start ETL batch consumer
+	l2Etl.tasks.Go(func() error {
```
Comment (hamdiallam) [inline indexer/etl/l2_etl.go:115]: I wonder if we should create an op-service wrapper that automatically handles panics here. This feels like copy/paste that could be DRYd out

File: indexer/indexer.go [hunk 4]
```diff
@@ -30,149 +30,228 @@ import (
...
+	if err := ix.initDB(ctx, cfg.DB); err != nil {
+		return fmt.Errorf("failed to init DB: %w", err)
+	}
+	if err := ix.initL1ETL(cfg.Chain); err != nil {
+		return fmt.Errorf("failed to init L1 ETL: %w", err)
+	}
+	if err := ix.initL2ETL(cfg.Chain); err != nil {
+		return fmt.Errorf("failed to init L2 ETL: %w", err)
+	}
+	if err := ix.initBridgeProcessor(cfg.Chain); err != nil {
```
Comment (hamdiallam) [inline indexer/indexer.go:156]: same here about having the routines put into `Start`. Like the earlier comment I could go either way though

**Review by hamdiallam — COMMENTED** [review]
This is awesome! Sick that we can make lifecycle more consistent

**Comment by hamdiallam** [comment]
I wonder if a "Service" can embed an struct defined in `op-service` that encapsulates some common things about a service like an abstract config struct, shutdown close context, and service logger? But maybe it's better to explicitly specify these fields though

File: indexer/api/api.go (hunk 2 above)
Comment (protolambda) [reply indexer/api/api.go:88]: I think we may still want to start the http server early if we want the address early on

File: indexer/etl/l2_etl.go (hunk 3 above)
Comment (protolambda) [reply indexer/etl/l2_etl.go:115]: done, we've a `tasks.Group` util now that handles panics in tasks

File: indexer/indexer.go (hunk 4 above)
Comment (protolambda) [reply indexer/indexer.go:156]: It makes sense, although I'm not keen to refactor even more within this same PR, it's already quite large. Maybe we can do this as a follow-up PR?

**Comment by protolambda** [comment]
>  encapsulates some common things about a service like an abstract config struct, shutdown close context, and service logger

So far we've tried to make op-service more compositional, rather than adding extensions to a single app in many different ways, which feels more like OOP type programming. I'm open to deduplicate logic where possible though, if it can stay compositional.

**Comment by protolambda** [comment]
Cleaned up the interrupt handling more: at the CLI root level we now add a context that has an attached interrupt handler. It's then up to the CLI action to cancel on interrupt, or use the full app lifecycle util. This way we can have different choices per sub-command; e.g. the DB migration command is still a simple cancel-on-interrupt, whereas the others are managed lifecycles.
And by retrieving the interrupt-blocker from the context this way, we can run the outer-most CLI app abstraction programmatically, and can still test the shutdown.

And made the DB connector dial respect the caller context, so it doesn't get stuck on a DB connection attempt when trying to shut down.

File: indexer/api/config.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	Closer          func() error
+}
+
+// DBConfigConnector implements a fully config based DBConnector
+type DBConfigConnector struct {
+	config.DBConfig
+}
+
+func (cfg *DBConfigConnector) OpenDB(ctx context.Context, log log.Logger) (*DB, error) {
+	// TODO: pass through the ctx argument, so we can interrupt while connecting
```
[the remark this answers is not in the document]
Comment (protolambda) [reply indexer/api/config.go]: fixed this

File: indexer/processors/bridge.go [hunk 6]
```diff
@@ -57,33 +65,66 @@ func NewBridgeProcessor(log log.Logger, db *database.DB, metrics bridge.Metricer
...
+				b.log.Info("stopping bridge processor")
+				return nil
+
+			// Tickers
+			case <-startup:
+			case <-l1EtlUpdates:
+			}
+
+			done := b.metrics.RecordInterval()
+			// TODO: why log all the errors and return the same thing, if we just return the error, and log here?
```
[the remark this answers is not in the document]
Comment (protolambda) [reply indexer/processors/bridge.go]: added github issue

**Review by hamdiallam — APPROVED** [review]
LGTM! this is awesome

**Comment by protolambda** [comment]
Will resolve the merge conflicts, one moment

**Comment by protolambda** [comment]
Merged in `develop` to resolve. #7985 was conflicting. The error handling on DB closing is already handled by this PR, and propagated properly. The write-timeout that was added to the middle-ware is new on develop, but was easy to merge.

**Comment by protolambda** [comment]
One indexer test still flakes: `TestE2EBridgeTransfersL2ToL1MessagePasserETHReceive`

```
    bridge_transfers_e2e_test.go:333: 
        	Error Trace:	/root/project/indexer/e2e_tests/bridge_transfers_e2e_test.go:333
        	Error:      	Received unexpected error:
        	            	failed to get receipt: context deadline exceeded
        	Test:       	TestE2EBridgeTransfersL2ToL1MessagePasserETHReceive
    config.go:43:               �[32mINFO �[0m[11-02|16:12:41.072] API service closed test DB view          �[32mrole�[0m=indexer_api
    api.go:111:                 �[32mINFO �[0m[11-02|16:12:41.072] API service shutdown complete            �[32mrole�[0m=indexer_api
    etl.go:65:                  �[32mINFO �[0m[11-02|16:12:41.072] stopped etl worker loop                  �[32mrole�[0m=indexer     �[32metl�[0m=l1
    l1_etl.go:140:              �[32mINFO �[0m[11-02|16:12:41.072] No more batches, shutting down L1 batch handler �[32mrole�[0m=indexer     �[32metl�[0m=l1
    etl.go:65:                  �[32mINFO �[0m[11-02|16:12:41.072] stopped etl worker loop                  �[32mrole�[0m=indexer     �[32metl�[0m=l2
    l2_etl.go:120:              �[32mINFO �[0m[11-02|16:12:41.072] No more batches, shutting down L2 batch handler �[32mrole�[0m=indexer     �[32metl�[0m=l2
    bridge.go:97:               �[32mINFO �[0m[11-02|16:12:41.072] stopping bridge processor                �[32mrole�[0m=indexer     �[32mprocessor�[0m=bridge
    db.go:107:                  �[32mINFO �[0m[11-02|16:12:41.072] closing database                         �[32mrole�[0m=indexer     �[32mmodule�[0m=db
    indexer.go:134:             �[32mINFO �[0m[11-02|16:12:41.072] indexer stopped                          �[32mrole�[0m=indexer
--- FAIL: TestE2EBridgeTransfersL2ToL1MessagePasserETHReceive (125.57s)
```
At least it appears to shutdown nicely on failure :sweat_smile: 

The WaitForReceipt function has a 2 minute timeout (applied to the context internally). Maybe that's too short?

**Review by ajsutton — APPROVED** [review]
Looks good. I don't know the indexer enough to be entirely confident in all those details (relying on Hamdi's review for that really) but the life cycle logic looks good and concurrency elements all check out as far as I can see.
